Parse discard binding

JLHwung opened the pull request:
<!--
Before making a PR, please read our contributing guidelines
https://github.com/babel/babel/blob/main/CONTRIBUTING.md

Please note that the Babel Team requires two approvals before merging most PRs.

For issue references: Add a comma-separated list of a [closing word](https://help.github.com/articles/closing-issues-via-commit-messages/) followed by the ticket number fixed by the PR. (it should be underlined in the preview if done correctly)

If you are making a change that should have a docs update: submit another PR to https://github.com/babel/website
-->

| Q                        | A <!--(Can use an emoji 👍) -->
| ------------------------ | ---
| Fixed Issues?            | Support [Discard Binding Proposal](https://tc39.es/proposal-discard-binding) in `@babel/parser` and `@babel/generator`
| Patch: Bug Fix?          |
| Major: Breaking Change?  |
| Minor: New Feature?      | Yes
| Tests Added + Pass?      | Yes
| Documentation PR Link    | <!-- If only readme change, add `[skip ci]` to your commits -->
| Any Dependency Changes?  |
| License                  | MIT

<!-- Describe your changes below in as much detail as possible -->
The AST design is aligned to the current ESTree proposal: https://github.com/estree/estree/blob/master/experimental/discard-binding.md

This PR will be merged to the umbrella branch `feat-discard-binding`.

JLHwung commented on packages/babel-parser/test/fixtures/experimental/discard-binding/using-variable-declarator/input.js:
```diff
@@ -0,0 +1,34 @@
+}
+{
+  // using declarator in for-of
+  for(using void of []);
```
Although the parser currently accepts this production, it is not allowed in the draft spec. I have filed an issue https://github.com/tc39/proposal-discard-binding/issues/14

nicolo-ribaudo replied:
Yeah this seems like a spec bug.

nicolo-ribaudo commented on packages/babel-plugin-transform-object-rest-spread/src/shouldStoreRHSInTemporaryVariable.ts:
```diff
@@ -12,7 +12,10 @@ export default function shouldStoreRHSInTemporaryVariable(
   if (!node) return false;
   if (node.type === "ArrayPattern") {
-    const nonNullElements = node.elements.filter(element => element !== null);
+    const nonNullElements = node.elements.filter(
```
Would it be possible to add a test for this change, using this plugin together with just syntax support for void patterns?

nicolo-ribaudo commented on packages/babel-types/src/definitions/core.ts:
```diff
@@ -1233,7 +1233,12 @@ defineType("VariableDeclarator", {
         !process.env.BABEL_8_BREAKING && !process.env.BABEL_TYPES_8_BREAKING
           ? assertNodeType("LVal")
-          : assertNodeType("Identifier", "ArrayPattern", "ObjectPattern"),
+          : assertNodeType(
```
In the `VariableDeclaration` builder, can we disallow `VoidPattern` for `var`/`let`/`const`?

JLHwung replied:
https://github.com/babel/babel/pull/17163/commits/3bdcdd3f82ab2648f1ca76f79a3f7a3051f38358

nicolo-ribaudo reviewed (commented):
I believe multiple delegates are still unhappy with the idea of using `void`. Could we maybe add a required `syntax: "void"` option to the parser plugin?

JLHwung commented:
> Could we maybe add a required syntax: "void" option to the parser plugin?

Sure, I will add a `syntaxType` option since we already have one for the record and tuple transform. 

> I believe multiple delegates are still unhappy with the idea of using void.

Out of curiosity, do you know if there are other type candidates? If `void` will not be likely adopted, then we should consider rename `VoidPattern` to `DiscardPattern` so that it is not tied to the `void` token.

nicolo-ribaudo commented:
> Out of curiosity, do you know if there are other type candidates? If `void` will not be likely adopted, then we should consider rename `VoidPattern` to `DiscardPattern` so that it is not tied to the `void` token.

The other possible approach is to relax the existing "duplicated declaration" errors in case where something is declared twice but not read, so that `{ let _ = 1; let _ = 2; }` would be valid. I think the current AST is fine, as in that case we would just use an `Identifier` node.

JLHwung commented:
Rebased. Now that #17217 has been merged, the last two validation harden commits is now replaced by https://github.com/babel/babel/pull/17163/commits/1a4fc60ec7c1246e9b0621f92138c1907e22baac.

nicolo-ribaudo commented on packages/babel-parser/src/parser/statement.ts:
```diff
@@ -351,7 +351,10 @@ export default abstract class StatementParser extends ExpressionParser {
       return false;
-    } else if (tokenIsIdentifier(type) && !this.hasFollowingLineBreak()) {
+    } else if (
+      (tokenIsIdentifier(type) || type === tt._void) &&
```
Probably a pointless micro-optimization, feel free to ignore. In `tt`s we currently have identifiers right after keywords. We could move `void` to the end of the keywords, so that we can have a single range covering `void` and identifiers.

JLHwung replied:
Good observation. There are a few other candidates that can benefit from this micro-optimization, such as `this` used by `tsSkipParameterStart` or `class` used by the `abstract` arm of the `tsParseDeclaration`, but we can at most pick one token unless there is demand for both `isAorBorIdentifier` and `isBorIdentifier`. 

Since discard-binding is still an early proposal and we might bikeshed the syntax type, I tend to leave it as-is. If it reaches stage 4 as-is and becomes very popular, we will optimize it then.

nicolo-ribaudo commented on packages/babel-parser/test/fixtures/experimental/discard-binding/array-pattern/output.json:
```diff
@@ -0,0 +1,519 @@
+  "type": "File",
+  "start":0,"end":566,"loc":{"start":{"line":1,"column":0,"index":0},"end":{"line":36,"column":1,"index":566}},
+  "errors": [
+    "SyntaxError: Binding invalid left-hand side in array destructuring pattern. (35:17)"
```
Is `try {} catch ([void]) {}` intended to be invalid?

JLHwung replied:
Good catch, this is a bug in our implementation.

nicolo-ribaudo commented on packages/babel-parser/test/fixtures/experimental/discard-binding/invalid-array-expression/output.json:
```diff
@@ -0,0 +1,230 @@
+  "type": "File",
+  "start":0,"end":108,"loc":{"start":{"line":1,"column":0,"index":0},"end":{"line":9,"column":21,"index":108}},
+  "errors": [
+    "SyntaxError: Unexpected void binding. (1:4)",
```
I would prefer something like `When not used in a binding position, "void" must be followed by an expression.`

JLHwung replied:
On a second thought, I extracted a new error type for the suggested message when CoverDiscardElement is seen. So that we still throw "unexpected void binding" on cases where `void` is indeed a binding but not allowed in the spec:
```js
var void = initializer;
```

https://github.com/babel/babel/pull/17163/commits/f8d042b492a55fc70e85e5b16cc56ddd96ac367a

nicolo-ribaudo commented:
@liuxingbaoyu Do you think you could review this in the next few days? It would be great to include it as merged in the STA report for April :)

liuxingbaoyu commented:
Of course! I will finish it soon. :)

nicolo-ribaudo commented:
(actually, I realized this will wait anyway because it's a new feature, but reviewing it will still unlock work on the transform!)

liuxingbaoyu commented on packages/babel-parser/src/parser/expression.ts:
```diff
@@ -3210,6 +3246,39 @@ export default abstract class ExpressionParser extends LValParser {
+    afterLeftParse?: Function,
+  ) {
+    if (refExpressionErrors != null && this.match(tt._void)) {
+      const nextType = this.lookahead().type;
+      if (nextType === tt.comma || nextType === close || nextType === tt.eq) {
```
```suggestion
      const nextCode = this.lookaheadCharCode();
      if (
        nextCode === charCodes.comma ||
        nextCode ===
          (close === tt.bracketR
            ? charCodes.rightSquareBracket
            : close === tt.braceR
              ? charCodes.rightCurlyBrace
              : charCodes.rightParenthesis) ||
        nextCode === charCodes.equalsTo
      ) {
```
This passes all the tests. Does this work?

nicolo-ribaudo commented:
@JLHwung As this PR is ready, let's merge it to the `feat-discard-binding` branch so that you can open the transform PR on top of it?
